shared/bundle_analysis/report.py

```diff
@@ -170,6 +171,104 @@ def ingest(self, path: str) -> int:
+        }
+        for curr_asset in curr_bundle_report.asset_reports():
+            if curr_asset.asset_type == models.AssetType.JAVASCRIPT:
+                if curr_asset.hashed_name in prev_asset_hashed_names:
+                    ret.add(
+                        (
+                            prev_asset_hashed_names[curr_asset.hashed_name],
+                            curr_asset.uuid,
+                        )
+                    )
+        return ret
+
+    def _associate_bundle_report_assets_by_module_names(
+        self, curr_bundle_report, prev_bundle_report
+    ) -> Set:
+        """
+        Rule 2
+        Returns a set of pairs of UUIDs (the current asset UUID and prev asset UUID)
+        representing that the curr asset UUID should be updated to the prev asset UUID
+        because there exists a prev asset where all its module names are the same as the
+        curr asset module names
+        """
+        ret = set()
+        prev_module_asset_mapping = {}
+        for prev_asset in prev_bundle_report.asset_reports():
+            if prev_asset.asset_type == models.AssetType.JAVASCRIPT:
+                prev_modules = tuple(
+                    sorted(frozenset([m.name for m in prev_asset.modules()]))
+                )
+                # NOTE: Assume two non-related assets CANNOT have the same set of modules
+                # though in reality there can be rare cases of this but we
+                # will deal with that later if it becomes a prevalent problem
+                prev_module_asset_mapping[prev_modules] = prev_asset.uuid
+
+        for curr_asset in curr_bundle_report.asset_reports():
+            if curr_asset.asset_type == models.AssetType.JAVASCRIPT:
+                curr_modules = tuple(
+                    sorted(frozenset([m.name for m in curr_asset.modules()]))
+                )
+                if curr_modules in prev_module_asset_mapping:
+                    ret.add(
+                        (
+                            prev_module_asset_mapping[curr_modules],
+                            curr_asset.uuid,
+                        )
+                    )
+        return ret
+
+    def associate_previous_assets(
+        self, prev_bundle_analysis_report: Any
+    ) -> "BundleAnalysisReport":
+        """
+        Only associate past asset if it is Javascript or Typescript types
+        and belonging to the same bundle name
+        Associated if one of the following is true
+        Rule 1. Previous and current asset have the same hashed name
+        Rule 2. Previous and current asset shared the same set of module names
+        """
+        for curr_bundle_report in self.bundle_reports():
+            for prev_bundle_report in prev_bundle_analysis_report.bundle_reports():
+                if curr_bundle_report.name == prev_bundle_report.name:
+                    associated_assets_found = set()
+
+                    # Rule 1 check
+                    associated_assets_found |= (
```

Comment:
Even the code is going "|=" (i.e. 😐)
(Jokes aside I do think this is an improvement, thanks)